src/datajson.c

```diff
@@ -0,0 +1,1009 @@
+static bool IsFloat(const char *in, size_t ins)
+{
+    char *endptr;
+    float val = strtof(in, &endptr);
+    const char *end_ins = in + ins - 1;
+    if (val != 0 && (endptr == end_ins)) {
+        return true;
+    }
+    /* if value is 0 then we need to check if some parsing has been done */
+    if (val == 0 && (endptr == in)) {
+        return false;
+    }
+    return true;
+}
+
+static int ParseJsonLine(const char *in, size_t ins, DataJsonType *rep_out)
+{
+    json_error_t jerror;
+    json_t *msg = json_loads(in, 0, &jerror);
+    if (msg == NULL) {
+        /* JANSSON does not see an integer, float or a string as valid JSON.
+           So we need to exclude them from failure. */
+        if (!IsFloat(in, ins) && !((in[0] == '"') && (in[ins - 1] == '"'))) {
+            SCLogWarning("dataset: Invalid json: %s: '%s'\n", jerror.text, in);
+            return -1;
+        }
+    } else {
+        json_decref(msg);
+    }
+    rep_out->len = ins;
+    rep_out->value = SCStrndup(in, ins);
+    if (rep_out->value == NULL) {
+        return -1;
+    }
+    return 0;
+}
+
+static json_t *GetSubObjectByKey(json_t *json, const char *key)
+{
+    if (!json || !key || !json_is_object(json)) {
+        return NULL;
+    }
+
+    const char *current_key = key;
+    json_t *current = json;
+    while (current_key) {
+        const char *dot = strchr(current_key, '.');
+
+        size_t key_len = dot ? (size_t)(dot - current_key) : strlen(current_key);
+        char key_buffer[key_len + 1];
+        strlcpy(key_buffer, current_key, key_len + 1);
+
+        if (json_is_object(current) == false) {
+            return NULL;
+        }
+        current = json_object_get(current, key_buffer);
+        if (current == NULL) {
+            return NULL;
+        }
+        current_key = dot ? dot + 1 : NULL;
+    }
+    return current;
+}
+
+static int ParseJsonFile(const char *file, json_t **array, char *key)
+{
+    json_t *json;
+    json_error_t error;
+    /* assume we have one single JSON element in FILE */
+    json = json_load_file(file, 0, &error);
+    if (json == NULL) {
+        FatalErrorOnInit("can't load JSON, error on line %d: %s", error.line, error.text);
+        return -1;
+    }
+
+    if (key == NULL || strlen(key) == 0) {
+        *array = json;
+    } else {
+        *array = GetSubObjectByKey(json, key);
+        if (*array == NULL) {
+            SCLogError("dataset: %s failed to get key '%s'", file, key);
+            json_decref(json);
+            return -1;
+        }
+        json_incref(*array);
+        json_decref(json);
+    }
+    if (!json_is_array(*array)) {
+        FatalErrorOnInit("not an array");
+        json_decref(*array);
+        return -1;
+    }
+    return 0;
+}
+
+/**
+ *  \retval 1 data was added to the hash
+ *  \retval 0 data was not added to the hash as it is already there
+ *  \retval -1 failed to add data to the hash
+ */
+static int DatajsonAddString(
```

Comment:
Looking at this it should just be in datasets.c as `DatasetAddStringwJSON`, similar to `DatasetAddStringwRep` I think

Reply:
The datasets.c file is really too long with more than 1500 lines and with a lot of functions having similar name it is complicated to navigate. Having a separate file is convenient for these reasons.